products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -25,96 +25,127 @@ using namespace llbuild::buildsystem;
 
 namespace {
 
-  class CAPIBuildDBKeysResult {
-  private:
-    const std::vector<KeyType> keys;
-    
-  public:
-    CAPIBuildDBKeysResult(const std::vector<KeyType>& keys): keys(keys) { }
-    
-    CAPIBuildDBKeysResult(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
-    CAPIBuildDBKeysResult& operator=(const CAPIBuildDBKeysResult&) LLBUILD_DELETED_FUNCTION;
-    
-    const size_t size() {
-      return keys.size();
+class CAPIBuildDB: public BuildDBDelegate {
+  /// The key table, for memory caching of key to key id mapping.
+  llvm::StringMap<KeyID> keyTable;
+  
+  /// The mutex that protects the key table.
+  std::mutex keyTableMutex;
+  
+  std::unique_ptr<BuildDB> _db;
+  
+  /// The mutex that protects the key cache.
+  std::mutex keyCacheMutex;
+  
+  std::unordered_map<KeyID, CAPIBuildKey *> keyCache;
+  
+  CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
+    _db = createSQLiteBuildDB(path, clientSchemaVersion, /* recreateUnmatchedVersion = */ false, error_out);
+  }
+  
+  bool fetchKeysIfNecessary(std::string *error) {
+    if (keyCache.size() > 0) {
+      return true;
     }
+    std::vector<KeyType> keys;
     
-    llb_build_key_t *keyAtIndex(int32_t index) {
-      auto buildKey = BuildKey::fromData(keys[index]);
-      return (llb_build_key_t *)new CAPIBuildKey(buildKey);
-    }
-  };
-
-  class CAPIBuildDB: public BuildDBDelegate {
-    /// The key table, for memory caching of key to key id mapping.
-    llvm::StringMap<KeyID> keyTable;
-    
-    /// The mutex that protects the key table.
-    std::mutex keyTableMutex;
-    
-    std::unique_ptr<BuildDB> _db;
+    auto success = _db->getKeys(keys, error);
     
-    CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
-      _db = createSQLiteBuildDB(path, clientSchemaVersion, /* recreateUnmatchedVersion = */ false, error_out);
+    if (!error->empty()) {
+      return success;
     }
     
-  public:
-    static CAPIBuildDB *create(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
-      auto databaseObject = new CAPIBuildDB(path, clientSchemaVersion, error_out);
-      if (databaseObject->_db == nullptr || !error_out->empty() || !databaseObject->buildStarted(error_out)) {
-        delete databaseObject;
-        return nullptr;
-      }
-      
-      // The delegate  caches keys for key ids
-      databaseObject->_db.get()->attachDelegate(databaseObject);
-      
-      return databaseObject;
-    }
+    buildUpKeyCache(keys);
     
-    bool isValid(std::string *error_out) {
-      if (_db == nullptr) {
-        return false;
-      }
-      
-      return buildStarted(error_out);
-    }
+    return success;
+  }
+  
+  void buildUpKeyCache(std::vector<KeyType> &allKeys) {
+    std::lock_guard<std::mutex> guard(keyCacheMutex);
```

Comment:
Hm, that's true. I don't expect that to happen but what would be the appropriate handling in that case? Ideally the first operation would cancel and the second one would be unlocked. Technically it might make sense to let the first finish, unlock the mutex and therefore unlock the second operation?

Reply:
I added a check if the `size()` of the cache and the given vector is equal and return early in that case. That makes it still possible to override it.